### PR TITLE
Testcases

### PR DESCRIPTION
**Pull request by mubaris**

### CODE REVIEW

File: contracts/receivers/aave-v2-receiver/helpers.sol [hunk 1]
```diff
@@ -49,7 +49,12 @@ abstract contract Helpers is Stores, DSMath, Variables {
         return data;
     }
 
-    function isPositionSafe() internal returns (bool isOk) {
+    function getCollateralStatus(address token, address user) internal view returns (bool enabled) {
+        (uint bal, , , , , , , , bool isCol) = aaveData.getUserReserveData(token, user);
+        enabled = bal > 0 && !isCol;
+    }
```
Comment (KaymasJain) [inline contracts/receivers/aave-v2-receiver/helpers.sol:51]: @thrilok209 where are you using this function?

File: contracts/senders/aave-v2-migrator/variables.sol [hunk 2]
```diff
@@ -29,28 +30,28 @@ contract Variables {
...
+    
+    /**
+    * @dev Polygon Receiver contract
+    */
+    address constant internal polygonReceiver = 0xA35f3FEFEcb5160327d1B6A210b60D1e1d7968e3; // This is fake addr. TODO @thrilok209: Replace this
+    
+    /**
+    * @dev Flashloan contract
+    */
+    FlashloanInterface constant internal flashloanContract = FlashloanInterface(0x709407B70ce0845996108205aA92aE87B22F7b1a); // TODO @thrilok209: Replace this
```
Comment (KaymasJain) [inline contracts/senders/aave-v2-migrator/variables.sol:48]: We have to replace these addresses. Right?

File: contracts/senders/aave-v2-migrator/variables.sol [hunk 3]
```diff
@@ -71,13 +72,39 @@ contract Variables {
...
     RootChainManagerInterface public constant rootChainManager = RootChainManagerInterface(0xA0c68C638235ee32657e8f720a23ceC1bFc77C77);
+    
+    
+    // Storage variables //
+    
+    /**
+    * @dev This will be used to have debt/collateral ratio always 20% less than liquidation
+    */
+    // TODO @KaymasJain: Is this number correct for it?
+    uint public safeRatioGap = 800000000000000000; // 20%?
```
Comment (KaymasJain) [inline contracts/senders/aave-v2-migrator/variables.sol:92]: Yes! It's correct. We're taking 100% of the position and multiplying it by 80%.

File: contracts/senders/aave-v2-migrator/variables.sol (hunk 2 above)
Comment (thrilok209) [reply contracts/senders/aave-v2-migrator/variables.sol:48]: Yes. I will replace before deployment.

File: contracts/senders/aave-v2-migrator/variables.sol [hunk 4]
```diff
@@ -71,13 +72,39 @@ contract Variables {
...
+
+    /**
+    * @dev fee on collateral
+    */
+    // TODO @KaymasJain: Is this number correct for it?
+    uint public fee = 998000000000000000; // 0.2% (99.8%) on collateral? TODO: Is this right?
+
```
Comment (KaymasJain) [inline contracts/senders/aave-v2-migrator/variables.sol:99]: Yep! 0.2% minus from 1e18

File: contracts/senders/aave-v2-migrator/variables.sol (hunk 3 above)
Comment (thrilok209) [reply contracts/senders/aave-v2-migrator/variables.sol:92]:
```suggestion
    uint public safeRatioGap = 800000000000000000; // 80%
```

File: contracts/senders/aave-v2-migrator/variables.sol (hunk 4 above)
Comment (thrilok209) [reply contracts/senders/aave-v2-migrator/variables.sol:99]:
```suggestion
    uint public fee = 998000000000000000; // 0.2% (99.8%) on collateral

```

File: contracts/liquidity.sol [hunk 5]
```diff
@@ -261,7 +262,7 @@ interface MigrationInterface {
 contract Setup {
     address public constant soloAddr = 0x1E0447b19BB6EcFdAe1e4AE1694b0C3659614e4e;
     address public constant wethAddr = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
-    MigrationInterface public constant migrationAddr = MigrationInterface(address(0)); // TODO: Migration address
+    MigrationInterface public constant migrationAddr = MigrationInterface(0xE3901a1bb3EefBF6bc90c201076e9068b646C1f3); // TODO: Migration address
```
Comment (thrilok209) [inline contracts/liquidity.sol:265]:
```suggestion
    MigrationInterface public constant migrationAddr = MigrationInterface(0xE3901a1bb3EefBF6bc90c201076e9068b646C1f3); // TODO @thrilok209: Chnage to Migration address
```

File: contracts/liquidity.sol [hunk 6]
```diff
@@ -271,6 +272,7 @@ contract Setup {
 contract DydxFlashloaner is Setup, ICallee, DydxFlashloanBase, DSMath {
     using SafeERC20 for IERC20;
 
+    // TODO: Change the migration call to .call so we can update the settle function later with flashloan
```
Comment (thrilok209) [inline contracts/liquidity.sol:275]:
```suggestion
```

File: contracts/liquidity.sol [hunk 7]
```diff
@@ -279,14 +280,14 @@ contract DydxFlashloaner is Setup, ICallee, DydxFlashloanBase, DSMath {
...
+        (bytes memory callData, uint ethAmt) = abi.decode(
             data,
-            (AaveDataRaw, address, uint)
+            (bytes, uint)
         );
 
         wethContract.transfer(address(migrationAddr), ethAmt);
 
-        migrationAddr.migrateFlashCallback(_data, dsa, ethAmt);
+        Address.functionCall(address(migrationAddr), callData); // Failing here?? `revert Address: low-level call failed`
```
Comment (thrilok209) [inline contracts/liquidity.sol:290]:
```suggestion
        Address.functionCall(address(migrationAddr), callData);
```

File: contracts/receivers/aave-v2-receiver/helpers.sol [hunk 8]
```diff
@@ -28,104 +45,93 @@ abstract contract Helpers is Stores, DSMath, Variables {
         return data;
     }
 
-    function isPositionSafe() internal returns (bool isOk) {
-        // TODO: Check the final position health
+    function getCollateralStatus(address token, address user) internal view returns (bool enabled) {
+        (uint bal, , , , , , , , bool isCol) = aaveData.getUserReserveData(token, user);
+        enabled = bal > 0 && !isCol;
+    }
```
Comment (thrilok209) [inline contracts/receivers/aave-v2-receiver/helpers.sol:51]:
```suggestion
```

File: contracts/receivers/aave-v2-receiver/interfaces.sol [hunk 9]
```diff
@@ -5,12 +5,12 @@ interface AccountInterface {
     function enable(address) external;
     function disable(address) external;
     function isAuth(address) external view returns (bool);
-    function cast(
+    function castMigrate(
```
Comment (thrilok209) [inline contracts/receivers/aave-v2-receiver/interfaces.sol:8]: @KaymasJain Here we are going to use `cast`, right?

File: contracts/senders/aave-v2-migrator/events.sol [hunk 10]
```diff
@@ -5,25 +5,41 @@ contract Events {
...
-        uint[] amts
+        uint256[] amts
     );
 
     event LogWithdraw(
         address owner,
         address[] tokens,
-        uint[] amts
+        uint256[] amts
+    );
```
Comment (thrilok209) [inline contracts/senders/aave-v2-migrator/events.sol:15]: @mubaris @KaymasJain We can remove these, right?

File: contracts/receivers/aave-v2-receiver/interfaces.sol (hunk 9 above)
Comment (KaymasJain) [reply contracts/receivers/aave-v2-receiver/interfaces.sol:8]: Yeah! We can use the normal `cast`. There shouldn't be any issue. Right?

File: contracts/senders/aave-v2-migrator/interfaces.sol [hunk 11]
```diff
@@ -1,6 +1,20 @@
...
+    //bit 48-55: Decimals
+    //bit 56: Reserve is active
+    //bit 57: reserve is frozen
+    //bit 58: borrowing is enabled
+    //bit 59: stable rate borrowing enabled
+    //bit 60-63: reserved
+    //bit 64-79: reserve factor
+    uint256 data;
+}
+
```
Comment (thrilok209) [inline contracts/senders/aave-v2-migrator/interfaces.sol:17]:
```suggestion
```

File: contracts/senders/aave-v2-migrator/interfaces.sol [hunk 12]
```diff
@@ -21,6 +35,10 @@ interface AaveInterface {
         uint256 ltv,
         uint256 healthFactor
     );
+    function getConfiguration(address asset)
+        external
+        view
+        returns (ReserveConfigurationMap memory);
```
Comment (thrilok209) [inline contracts/senders/aave-v2-migrator/interfaces.sol:41]:
```suggestion
    function getConfiguration(address asset)
        external
        view
        returns (ReserveConfigurationMap memory);
```

File: contracts/receivers/aave-v2-receiver/interfaces.sol (hunk 9 above)
Comment (KaymasJain) [reply contracts/receivers/aave-v2-receiver/interfaces.sol:8]: Or else we can change it to `castMigrate` with some restriction.

File: contracts/senders/aave-v2-migrator/main.sol [hunk 13]
```diff
@@ -7,26 +7,30 @@ import { TokenInterface } from "../../common/interfaces.sol";
 import { Helpers } from "./helpers.sol";
 import { AaveInterface, ATokenInterface, IndexInterface } from "./interfaces.sol";
 import { Events } from "./events.sol";
+import "hardhat/console.sol";
```
Comment (thrilok209) [inline contracts/senders/aave-v2-migrator/main.sol:10]:
```suggestion
```

File: contracts/senders/aave-v2-migrator/events.sol (hunk 10 above)
Comment (KaymasJain) [reply contracts/senders/aave-v2-migrator/events.sol:15]: Yep!

File: contracts/senders/aave-v2-migrator/main.sol [hunk 14]
```diff
@@ -214,13 +154,10 @@ contract MigrateResolver is LiquidityResolver {
...
         _checkRatio(data);
 
-        if (ethAmt > 0) {
-            aave.withdraw(wethAddr, ethAmt, address(this));
-        }
-
         isPositionSafe();
 
         stateSender.syncState(polygonReceiver, abi.encode(data));
+        console.logBytes(abi.encode(data));
```
Comment (thrilok209) [inline contracts/senders/aave-v2-migrator/main.sol:160]:
```suggestion
```

File: contracts/senders/aave-v2-migrator/main.sol [hunk 15]
```diff
@@ -154,22 +77,26 @@ contract LiquidityResolver is Helpers, Events {
...
                     aave.repay(_token, supplyBal, 2, address(this));
                 }
             }
         }
         for (uint i = 0; i < _tokens.length; i++) {
-            aave.withdraw(_tokens[i], _amts[i], address(this));
-            // TODO: transfer to polygon's receiver address "polygonReceiver"
+            address _token = _tokens[i] == ethAddr ? wethAddr : _tokens[i];
+            aave.withdraw(_token, _amts[i], address(this));
+            // TODO: Verify this
```
Comment (thrilok209) [inline contracts/senders/aave-v2-migrator/main.sol:92]:
```suggestion
```

File: contracts/senders/aave-v2-migrator/interfaces.sol (hunk 11 above)
Comment (KaymasJain) [reply contracts/senders/aave-v2-migrator/interfaces.sol:17]: What is this?

**Review by thrilok209 — APPROVED** [review]

File: contracts/senders/aave-v2-migrator/interfaces.sol (hunk 11 above)
Comment (thrilok209) [reply contracts/senders/aave-v2-migrator/interfaces.sol:17]: Bit map of the different parameters stored for an asset. We can remove this as we are not using this anywhere

**Pull request merged**